File: src/main/java/akostenko/aicars/math/Decart.java
package akostenko.aicars.math;

import static java.lang.Math.PI;
import static java.lang.StrictMath.atan;
import static java.lang.StrictMath.sqrt;

public class Decart implements Vector {
    public static final Decart ZERO = new Decart(0,0);


    public final double x;
    public final double y;

    public Decart(double x, double y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public Polar toPolar() {
        double d;
        if (x != 0) {
            d = atan(y / x);
            if (d < 0) {
                d += PI;
            } else if (y == 0 && x < 0) {
                d += PI;
            }
            if (y < 0) {
                d += PI;
            }
        } else if (y != 0) {
            d = y > 0 ? PI/2 : 3*PI/2;
        } else {
            d = 0;
        }
        return new Polar(sqrt(x*x + y*y), d);
    }

    @Override
    public Decart toDecart() {
        return this;
    }

    @Override
    public Decart plus(Vector v) {
        return addDecart(v.toDecart());
    }

    Decart addDecart(Decart v) {
        return new Decart(x+ v.x, y + v.y);
    }

    @Override
    public Decart minus(Vector v) {
        return plus(v.negative());
    }

    @Override
    public Decart negative() {
        return multi(-1);
    }

    @Override
    public Decart multi(double k) {
        return new Decart(x * k, y * k);
    }

    @Override
    public Vector div(double k) {
        return multi(1./k);
    }

    @Override
    public Decart rotate(double radians) {
        return toPolar().rotate(radians).toDecart();
    }

    @Override
    public double module() {
        return sqrt(moduleSqr());
    }

    @Override
    public double moduleSqr() {
        return x*x + y*y;
    }

    @Override
    public double dot(Vector v) {
        return toPolar().dot(v.toPolar());
    }

    @Override
    public double cross(Vector v) {
        return toPolar().cross(v.toPolar());
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;

        Decart decart;
        if (o instanceof Vector) {
            decart = ((Vector) o).toDecart();
        } else return false;

        return (decart.x - x < PRECISION) && (decart.y - y < PRECISION);
    }

    @Override
    public int hashCode() {
        int result;
        long temp;
        temp = Double.doubleToLongBits(x);
        result = (int) (temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(y);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        return result;
    }

    @Override
    public String toString() {
        return "Decart{" +
                "x=" + x +
                ", y=" + y +
                '}';
    }
}

File: src/main/kotlin/akostenko/aicars/math/Decart.kt
package akostenko.aicars.math

import java.lang.StrictMath.PI
import java.lang.StrictMath.abs
import java.lang.StrictMath.atan
import java.lang.StrictMath.hypot
import java.lang.StrictMath.pow

class Decart(val x: Double, val y: Double) : Vector {

    private val polar : Polar by lazy {
        var d: Double
        if (x != 0.0) {
            d = atan(y / x)
            if (d < 0) {
                d += PI
            } else if (y == 0.0 && x < 0) {
                d += PI
            }
            if (y < 0) {
                d += PI
            }
        } else if (y != 0.0) {
            d = if (y > 0) PI / 2 else 3 * PI / 2
        } else {
            d = 0.0
        }
        Polar(hypot(x, y), d)
    }

    override fun toPolar(): Polar = polar

    override fun toDecart() : Decart = this

    override fun plus(v: Vector) : Decart = addDecart(v.toDecart())

    internal fun addDecart(v: Decart) : Decart = Decart(x + v.x, y + v.y)

    override fun minus(v: Vector) : Decart = plus(-v)

    override fun unaryMinus(): Decart = times(-1.0)

    override fun times(k: Double): Decart = Decart(x * k, y * k)

    override fun div(k: Double): Decart = times(1.0 / k)

    override fun rotate(radians: Double): Decart = toPolar().rotate(radians).toDecart()

    override fun module(): Double = hypot(x, y)

    override fun moduleSqr(): Double = pow(module(), 2.0)

    override fun dot(v: Vector): Double = toPolar().dot(v.toPolar())

    override fun cross(v: Vector): Double = toPolar().cross(v.toPolar())

    override fun equals(other: Any?): Boolean {
        if (this === other) return true

        if (other is Vector) {
            val otherDecart = other.toDecart()
            return abs(otherDecart.x - x) < Vector.PRECISION && abs(otherDecart.y - y) < Vector.PRECISION
        } else return false
    }

    override fun hashCode(): Int {
        var result = x.hashCode()
        result = 31 * result + y.hashCode()
        return result
    }

    override fun toString() : String = "Decart(x=$x, y=$y)"

    companion object {
        val ZERO = Decart(0.0, 0.0)
    }
}
